File: interaccion/src/reloj_nodo.cpp
/*Este nodo llamado reloj_nodo muestra la hora UTC, el tiempo transcurrido desde
el ultimo mensaje y envia una señal al dialogo_nodo confirmandole que sigue "vivo"
*/

//Se incluyen las cabeceras
#include "ros/ros.h"
#include "std_msgs/String.h"
#include "std_msgs/Bool.h"
#include <ctime>
#include "boost/date_time/posix_time/posix_time.hpp"

//Declaracion de macros auxiliares
#define RELOJ_MSG_NAME "still_alive"
#define COUNTDOWN_TIME 60


//Declaracion de namespaces
using namespace boost::posix_time;
using namespace ros;
using namespace std;

//Declaracion de variables globales
Time startTime;
bool clock_start = false;
int totalSeconds = 0;

Publisher publicadorTiempo; //Publicador de mensajes que envía still_alive

/**
* Esta función confirma el start recibido desde el nodo_dialogo
*/
void funcionCallback(const std_msgs::String::ConstPtr& msg){
 ROS_INFO("He recibido un mensaje con la informacion: %s", msg->data.c_str());
 startTime = Time::now();
 clock_start = true;
}

//Esta funcion confirma el reset recibido desde el dialogo nodo
void funcionCallback2(const std_msgs::String::ConstPtr& msg){
 ROS_INFO("He recibido un mensaje con la informacion: %s", msg->data.c_str());
 startTime = Time::now();
 clock_start = true;
}

/*timerCallback se ejecuta cada vez que se vence el tiempo del timer
*/
void timerCallback(const ros::TimerEvent&){
	std_msgs::Bool still_alive;
	still_alive.data = true;
	publicadorTiempo.publish(still_alive);
}

/*printClock se encarga de ejecutar la hora UTC
  y de manera adicionar imprime el tiempo transcurrido entre start y reset
*/
void printClock() {
	totalSeconds = (Time::now() - startTime).toSec();	//Segundos desde el inicio hasta el momento actual

	ptime t_local(second_clock::local_time());			//Declaracion de la variable de tiempo local
	ptime t_utc(second_clock::universal_time());		//Declaracion de la variable de tiempo UTC

	ROS_INFO("LOCAL HOUR: %s", to_simple_string(t_local).c_str());
	ROS_INFO("UTC HOUR: %s", to_simple_string(t_utc).c_str());

	ROS_INFO("SECONDS FROM START/RESET: %lf", (double)(Time::now()-startTime).toSec());
}

//funcion principal
int main(int argc, char **argv){
 //registra el nombre del nodo: reloj_nodo
 ros::init(argc, argv, "reloj_nodo");
 ros::NodeHandle nodoReloj;
 ROS_INFO("reloj_nodo creado y registrado"); //muestra en pantalla

 //si recibimos el mensaje cuyo topic es: "start_topic" llamamos a la función manejadora: funcionCallback
 ros::Subscriber subcriptor1 = nodoReloj.subscribe("start_topic", 0, funcionCallback);

  //si recibimos el mensaje cuyo topic es: "reset_topic" llamamos a la función manejadora: funcionCallback2
 ros::Subscriber subscriptor2 = nodoReloj.subscribe("reset_topic", 0, funcionCallback2);
 
 //Timer que se vence cada 60s para indicar que el nodo sigue activo
 Timer timer = nodoReloj.createTimer(Duration(COUNTDOWN_TIME), timerCallback);
 
 //Publicador de mensajes para indicar que el nodo sigue activo
 publicadorTiempo = nodoReloj.advertise<std_msgs::Bool>(RELOJ_MSG_NAME,0);

 //Frecuencia con la que se ejecuta el bucle principal
 ros::Duration seconds_sleep(3);

  while (ros::ok()){
    //Cuando se envie el primer mensaje, se comienza a llamar a printclock
  	if(clock_start){
  		printClock();
    }

    spinOnce();

    ROS_DEBUG("El nodo se duerme durante tres segundos");

    //rate.sleep para cumplir la frecuencia de tres segundos
    seconds_sleep.sleep();
 }
 return 0;
}

File: interaccion/src/reloj_nodo.py
import rospy
from std_msgs.msg import String,Bool
from datetime import datetime

#CALLBACK CUANDO RECIBE UN TOPIC DE RESET
def callbackR(data):
    #DEFINIMOS ACT COMO GLOBAL
    global act
    #DEFINIMOS ACT COMO EL TIEMPO EN EL MOMENTO QUE SE LLAMA AL CALLBACK
    act = rospy.get_rostime().secs

#CALLBACK CUANDO RECIBE UN TOPIC DE START
def callbackS(data):
    #DEFINIMOS ACT COMO GLOBAL
    global act

    #DEFINIMOS ACT COMO EL TIEMPO EN EL MOMENTO QUE SE LLAMA AL CALLBACK
    act = rospy.get_rostime().secs

    #COMENZAMOS A IMPRIMIR POR PANTALLA
    imprime()

#IMPRIME POR PANTALLA LOS CAMPOS HORA LOCAL, HORA UTC Y TIEMPO DESDE EL ULTIMO RESET/START
def imprime():
    #DEFINIMOS ACT COMO GLOBAL
    global act
    #EL RATE AL QUE QUEREMOS IMPRIMIR POR PANTALLA, EN ESTE CASO 3HZ
    r=rospy.Rate(3)
    #COMENZAMOS A IMPRIMIR
    while True:
        #EL TIEMPO ACTUAL
        now = rospy.get_rostime()
        #CONVERSION DE TIMESTAMP A UTC DE LA HORA UTC Y LA ACTUAL
        horautc=datetime.utcfromtimestamp(now.secs).strftime('%Hh:%Mm:%Ss')
        hora=datetime.utcfromtimestamp(now.secs+3600).strftime('%Hh:%Mm:%Ss')
        #IMPRIMIMOS
        rospy.loginfo("La hora UTC es: %s     La hora LOCAL es: %s    Segundos desde el ultimo reset/start es: %i" , horautc,hora,now.secs-act)
        r.sleep()

#CALLBACK CUANDO RECIBE EL STILL_ALIVE
def still_alive(event):
    #DEFINIMOS EL PUBLICADOR
    global pub
    #PUBLICAMOS
    pub.publish(True)

#FUNCION PRINCIPAL
def reloj():
    #DEFINIMOS EL PUBLICADOR COMO GLOBAL
    global pub
    #INICIAMOS EL NODO
    rospy.init_node('reloj', anonymous=True)
    #DEFINIMOS EL PUBLICADOR
    pub = rospy.Publisher('still_alive', Bool,queue_size=1)
    #DEFINIMOS LOS SUBSCRIPTORES
    rospy.Subscriber('reset_topic', String, callbackR)
    rospy.Subscriber('start_topic', String, callbackS)
    #DEFINIMOS EL TIMER QUE LANZA LA FUNCION STILL_ALIVE CADA 60S
    rospy.Timer(rospy.Duration(60), still_alive)
    #LO PONEMOS A CORRER
    rospy.spin()

if __name__=='__main__':
    #LO INICIALIZAMOS 
    reloj()
